saw-remote-api/src/SAWServer/JVMCrucibleSetup.hs

```diff
@@ -0,0 +1,197 @@
+      getTypedTerm env p >>= jvm_postcond
+
+    opts = defaultOptions
+    save name val = modify' (\(env, cenv) -> (Map.insert name val env, cenv))
+
+    getSetupVal ::
+      (Map ServerName ServerSetupVal, CryptolEnv) ->
+      CrucibleSetupVal (P.Expr P.PName) ->
+      JVMSetupM (MS.SetupValue CJ.JVM)
+    getSetupVal _ NullValue = JVMSetupM $ return $ MS.SetupNull ()
+                              {-
+    getSetupVal env (ArrayValue elts) =
+      do elts' <- mapM (getSetupVal env) elts
+         JVMSetupM $ return $ MS.SetupArray () elts'
+    getSetupVal env (FieldLValue base fld) =
+      do base' <- getSetupVal env base
+         JVMSetupM $ return $ MS.SetupField () base' fld
+    getSetupVal env (ElementLValue base idx) =
+      do base' <- getSetupVal env base
+         JVMSetupM $ return $ MS.SetupElem () base' idx
+    getSetupVal _ (GlobalInitializer name) =
+      JVMSetupM $ return $ MS.SetupGlobalInitializer () name
+    getSetupVal _ (GlobalLValue name) =
+      JVMSetupM $ return $ MS.SetupGlobal () name
+         -}
+    getSetupVal (env, _) (ServerValue n) = JVMSetupM $
+      resolve env n >>=
+      \case
+        Val x -> return x -- TODO add cases for the server values that
+                          -- are not coming from the setup monad
+                          -- (e.g. surrounding context)
+    getSetupVal (_, cenv) (CryptolExpr expr) = JVMSetupM $
+      do res <- liftIO $ getTypedTermOfCExp fileReader (biSharedContext bic) cenv expr
+         -- TODO: add warnings (snd res)
+         case fst res of
+           Right (t, _) -> return (MS.SetupTerm t)
+           Left err -> error $ "Cryptol error: " ++ show err -- TODO: report properly
+    getSetupVal _ _sv = error $ "unrecognized setup value" -- ++ show sv
+
+    getTypedTerm ::
+      (Map ServerName ServerSetupVal, CryptolEnv) ->
+      P.Expr P.PName ->
+      JVMSetupM TypedTerm
+    getTypedTerm (_, cenv) expr = JVMSetupM $
+      do res <- liftIO $ getTypedTermOfCExp fileReader (biSharedContext bic) cenv expr
+         -- TODO: add warnings (snd res)
+         case fst res of
+           Right (t, _) -> return t
+           Left err -> error $ "Cryptol error: " ++ show err -- TODO: report properly
+
+    resolve env name =
+       case Map.lookup name env of
+         Just v -> return v
+         Nothing -> error "Server value not found - impossible!" -- rule out elsewhere
+
+data JVMLoadClassParams
+  = JVMLoadClassParams ServerName String
+
+instance FromJSON JVMLoadClassParams where
+  parseJSON =
+    withObject "params for \"SAW/JVM/load class\"" $ \o ->
+    JVMLoadClassParams <$> o .: "name" <*> o .: "class"
+
+jvmLoadClass :: JVMLoadClassParams -> Method SAWState OK
+jvmLoadClass (JVMLoadClassParams serverName cname) =
+  do tasks <- view sawTask <$> getState
+     case tasks of
+       (_:_) -> raise $ notAtTopLevel $ map fst tasks
+       [] ->
+         do bic <- view sawBIC <$> getState
+            c <- tl $ loadJavaClass bic cname
```

Comment:
Since #82, `loadJavaClass` should be using the IO action returned from `getFileReader` so that the caching support will be transparent.

Reply:
Yeah, it should, but I believe it depends on code that hasn't been refactored to use that mechanism yet. The Java verification commands aren't functional yet, in other ways, so I'm inclined to hold off on implementing that. I added a comment to #15 to track this.

Reply:
Sounds reasonable to me.